In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.simplefilter('ignore')

In [2]:
df=pd.read_csv('heart_disease_data.csv.zip')
df.head()

Year LocationAbbr   LocationDesc DataSource PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare          None          None   
1  2005           US  United States   Medicare          None          None   
2  2007           US  United States   Medicare          None          None   
3  2008           US  United States   Medicare          None          None   
4  2004           US  United States   Medicare          None          None   

  PriorityArea3 PriorityArea4                 Category  \
0          None          None  Cardiovascular Diseases   
1          None          None  Cardiovascular Diseases   
2          None          None  Cardiovascular Diseases   
3          None          None  Cardiovascular Diseases   
4          None          None  Cardiovascular Diseases   

                    Topic  ... Break_Out_Category Break_Out CategoryId  \
0           Heart Failure  ...               Race     Other         C1   
1           Heart Failure  ...               Race     Other         C1   
2  Coronary Heart Disease  ...                Age       65+         C1   
3  Coronary Heart Disease  ...             Gender    Female         C1   
4           Heart Failure  ...            Overall   Overall         C1   

   TopicId  IndicatorID Data_Value_TypeID BreakOutCategoryId  BreakOutId  \
0       T5        MD501             Crude              BOC04       RAC07   
1       T5        MD501             Crude              BOC04       RAC07   
2       T4        MD301             Crude              BOC03       AGE06   
3       T4        MD301             Crude              BOC02       GEN02   
4       T5        MD501             Crude              BOC01       OVR01   

   LocationID GeoLocation  
0          59         NaN  
1          59         NaN  
2          59         NaN  
3          59         NaN  
4          59         NaN  

[5 rows x 29 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42640 entries, 0 to 42639
Data columns (total 29 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Year                        42640 non-null  int64  
 1   LocationAbbr                42640 non-null  object 
 2   LocationDesc                42640 non-null  object 
 3   DataSource                  42640 non-null  object 
 4   PriorityArea1               42640 non-null  object 
 5   PriorityArea2               42640 non-null  object 
 6   PriorityArea3               42640 non-null  object 
 7   PriorityArea4               42640 non-null  object 
 8   Category                    42640 non-null  object 
 9   Topic                       42640 non-null  object 
 10  Indicator                   42640 non-null  object 
 11  Data_Value_Type             42640 non-null  object 
 12  Data_Value_Unit             42640 non-null  object 
 13  Data_Value                  421

In [4]:
df.isnull().sum()

Year                              0
LocationAbbr                      0
LocationDesc                      0
DataSource                        0
PriorityArea1                     0
PriorityArea2                     0
PriorityArea3                     0
PriorityArea4                     0
Category                          0
Topic                             0
Indicator                         0
Data_Value_Type                   0
Data_Value_Unit                   0
Data_Value                      529
Data_Value_Alt                    0
Data_Value_Footnote_Symbol    42111
Data_Value_Footnote           42111
LowConfidenceLimit              529
HighConfidenceLimit             529
Break_Out_Category                0
Break_Out                         0
CategoryId                        0
TopicId                           0
IndicatorID                       0
Data_Value_TypeID                 0
BreakOutCategoryId                0
BreakOutId                        0
LocationID                  

In [5]:
df.shape

(42640, 29)

In [6]:
df=df.drop(['Data_Value_Footnote_Symbol','Data_Value_Footnote'],axis=1)
df.head(10)

Year LocationAbbr   LocationDesc DataSource   PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare            None          None   
1  2005           US  United States   Medicare            None          None   
2  2007           US  United States   Medicare            None          None   
3  2008           US  United States   Medicare            None          None   
4  2004           US  United States   Medicare            None          None   
5  2012           US  United States   Medicare            None          None   
6  2005           NY       New York   Medicare  Million Hearts          None   
7  2008           CT    Connecticut   Medicare  Million Hearts          None   
8  2008           ME          Maine   Medicare  Million Hearts          None   
9  2010           OH           Ohio   Medicare  Million Hearts          None   

  PriorityArea3 PriorityArea4                 Category  \
0          None          None  Cardiovascular Diseases   
1          None          None  Cardiovascular Diseases   
2          None          None  Cardiovascular Diseases   
3          None          None  Cardiovascular Diseases   
4          None          None  Cardiovascular Diseases   
5          None          None  Cardiovascular Diseases   
6          None          None  Cardiovascular Diseases   
7          None          None  Cardiovascular Diseases   
8          None          None  Cardiovascular Diseases   
9          None          None  Cardiovascular Diseases   

                    Topic  ... Break_Out_Category Break_Out CategoryId  \
0           Heart Failure  ...               Race     Other         C1   
1           Heart Failure  ...               Race     Other         C1   
2  Coronary Heart Disease  ...                Age       65+         C1   
3  Coronary Heart Disease  ...             Gender    Female         C1   
4           Heart Failure  ...            Overall   Overall         C1   
5           Heart Failure  ...               Race     Other         C1   
6                  Stroke  ...                Age       65+         C1   
7                  Stroke  ...                Age       75+         C1   
8                  Stroke  ...               Race     Other         C1   
9                  Stroke  ...                Age       75+         C1   

   TopicId  IndicatorID  Data_Value_TypeID  BreakOutCategoryId BreakOutId  \
0       T5        MD501              Crude               BOC04      RAC07   
1       T5        MD501              Crude               BOC04      RAC07   
2       T4        MD301              Crude               BOC03      AGE06   
3       T4        MD301              Crude               BOC02      GEN02   
4       T5        MD501              Crude               BOC01      OVR01   
5       T5        MD501              Crude               BOC04      RAC07   
6       T6        MD601              Crude               BOC03      AGE06   
7       T6        MD601              Crude               BOC03      AGE08   
8       T6        MD601              Crude               BOC04      RAC07   
9       T6        MD601              Crude               BOC03      AGE08   

  LocationID                               GeoLocation  
0         59                                       NaN  
1         59                                       NaN  
2         59                                       NaN  
3         59                                       NaN  
4         59                                       NaN  
5         59                                       NaN  
6         36   (42.82700103200045, -75.54397042699964)  
7          9   (41.56266102000046, -72.64984095199964)  
8         23  (45.254228894000505, -68.98503133599962)  
9         39   (40.06021014100048, -82.40426005599966)  

[10 rows x 27 columns]

In [7]:
df.shape

(42640, 27)

In [8]:
df['LowConfidenceLimit']

0        5.1
1        5.1
2        8.4
3        5.1
4        6.1
        ... 
42635    3.6
42636    3.3
42637    NaN
42638    4.3
42639    3.8
Name: LowConfidenceLimit, Length: 42640, dtype: float64

In [9]:
from sklearn.impute import SimpleImputer
median_imputer=SimpleImputer(strategy='median')

df[['Data_Value','LowConfidenceLimit','HighConfidenceLimit']]=median_imputer.fit_transform(df[['Data_Value','LowConfidenceLimit','HighConfidenceLimit']])
df.head()

Year LocationAbbr   LocationDesc DataSource PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare          None          None   
1  2005           US  United States   Medicare          None          None   
2  2007           US  United States   Medicare          None          None   
3  2008           US  United States   Medicare          None          None   
4  2004           US  United States   Medicare          None          None   

  PriorityArea3 PriorityArea4                 Category  \
0          None          None  Cardiovascular Diseases   
1          None          None  Cardiovascular Diseases   
2          None          None  Cardiovascular Diseases   
3          None          None  Cardiovascular Diseases   
4          None          None  Cardiovascular Diseases   

                    Topic  ... Break_Out_Category Break_Out CategoryId  \
0           Heart Failure  ...               Race     Other         C1   
1           Heart Failure  ...               Race     Other         C1   
2  Coronary Heart Disease  ...                Age       65+         C1   
3  Coronary Heart Disease  ...             Gender    Female         C1   
4           Heart Failure  ...            Overall   Overall         C1   

   TopicId  IndicatorID  Data_Value_TypeID  BreakOutCategoryId BreakOutId  \
0       T5        MD501              Crude               BOC04      RAC07   
1       T5        MD501              Crude               BOC04      RAC07   
2       T4        MD301              Crude               BOC03      AGE06   
3       T4        MD301              Crude               BOC02      GEN02   
4       T5        MD501              Crude               BOC01      OVR01   

  LocationID GeoLocation  
0         59         NaN  
1         59         NaN  
2         59         NaN  
3         59         NaN  
4         59         NaN  

[5 rows x 27 columns]

In [10]:
df.isnull().sum()

Year                     0
LocationAbbr             0
LocationDesc             0
DataSource               0
PriorityArea1            0
PriorityArea2            0
PriorityArea3            0
PriorityArea4            0
Category                 0
Topic                    0
Indicator                0
Data_Value_Type          0
Data_Value_Unit          0
Data_Value               0
Data_Value_Alt           0
LowConfidenceLimit       0
HighConfidenceLimit      0
Break_Out_Category       0
Break_Out                0
CategoryId               0
TopicId                  0
IndicatorID              0
Data_Value_TypeID        0
BreakOutCategoryId       0
BreakOutId               0
LocationID               0
GeoLocation            820
dtype: int64

In [11]:
df.isnull().sum()

Year                     0
LocationAbbr             0
LocationDesc             0
DataSource               0
PriorityArea1            0
PriorityArea2            0
PriorityArea3            0
PriorityArea4            0
Category                 0
Topic                    0
Indicator                0
Data_Value_Type          0
Data_Value_Unit          0
Data_Value               0
Data_Value_Alt           0
LowConfidenceLimit       0
HighConfidenceLimit      0
Break_Out_Category       0
Break_Out                0
CategoryId               0
TopicId                  0
IndicatorID              0
Data_Value_TypeID        0
BreakOutCategoryId       0
BreakOutId               0
LocationID               0
GeoLocation            820
dtype: int64

In [12]:
df['Topic'].value_counts()

Heart Failure                                 19240
Coronary Heart Disease                         4680
Stroke                                         4680
Major Cardiovascular Disease                   4680
Diseases of the Heart (Heart Disease)          4680
Acute Myocardial Infarction (Heart Attack)     4680
Name: Topic, dtype: int64

In [13]:
df=df.drop(['GeoLocation'],axis=1)

In [14]:
df.head()

Year LocationAbbr   LocationDesc DataSource PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare          None          None   
1  2005           US  United States   Medicare          None          None   
2  2007           US  United States   Medicare          None          None   
3  2008           US  United States   Medicare          None          None   
4  2004           US  United States   Medicare          None          None   

  PriorityArea3 PriorityArea4                 Category  \
0          None          None  Cardiovascular Diseases   
1          None          None  Cardiovascular Diseases   
2          None          None  Cardiovascular Diseases   
3          None          None  Cardiovascular Diseases   
4          None          None  Cardiovascular Diseases   

                    Topic  ... HighConfidenceLimit Break_Out_Category  \
0           Heart Failure  ...                 5.3               Race   
1           Heart Failure  ...                 5.2               Race   
2  Coronary Heart Disease  ...                 8.5                Age   
3  Coronary Heart Disease  ...                 5.1             Gender   
4           Heart Failure  ...                 6.1            Overall   

  Break_Out  CategoryId  TopicId  IndicatorID  Data_Value_TypeID  \
0     Other          C1       T5        MD501              Crude   
1     Other          C1       T5        MD501              Crude   
2       65+          C1       T4        MD301              Crude   
3    Female          C1       T4        MD301              Crude   
4   Overall          C1       T5        MD501              Crude   

  BreakOutCategoryId BreakOutId LocationID  
0              BOC04      RAC07         59  
1              BOC04      RAC07         59  
2              BOC03      AGE06         59  
3              BOC02      GEN02         59  
4              BOC01      OVR01         59  

[5 rows x 26 columns]

In [15]:
df['Category'].value_counts()

Cardiovascular Diseases    42640
Name: Category, dtype: int64

In [16]:
df.isnull().sum()

Year                   0
LocationAbbr           0
LocationDesc           0
DataSource             0
PriorityArea1          0
PriorityArea2          0
PriorityArea3          0
PriorityArea4          0
Category               0
Topic                  0
Indicator              0
Data_Value_Type        0
Data_Value_Unit        0
Data_Value             0
Data_Value_Alt         0
LowConfidenceLimit     0
HighConfidenceLimit    0
Break_Out_Category     0
Break_Out              0
CategoryId             0
TopicId                0
IndicatorID            0
Data_Value_TypeID      0
BreakOutCategoryId     0
BreakOutId             0
LocationID             0
dtype: int64

In [17]:
from sklearn.preprocessing import LabelEncoder
lab=LabelEncoder()
df['Topic']=lab.fit_transform(df['Topic'])
df['Break_Out_Category']=lab.fit_transform(df['Break_Out_Category'])
df['Break_Out_CategoryId']=lab.fit_transform(df['BreakOutCategoryId'])
df['BreakOutId']=lab.fit_transform(df['BreakOutId'])
df['TopicId']=lab.fit_transform(df['TopicId'])
df['Data_Value_Unit']=lab.fit_transform(df['Data_Value_Unit'])
df['PriorityArea1']=lab.fit_transform(df['PriorityArea1'])
df['PriorityArea3']=lab.fit_transform(df['PriorityArea3'])


In [18]:
df['Topic'].value_counts()

3    19240
1     4680
5     4680
4     4680
2     4680
0     4680
Name: Topic, dtype: int64

In [19]:
df['Topic'].value_counts()

3    19240
1     4680
5     4680
4     4680
2     4680
0     4680
Name: Topic, dtype: int64

In [20]:
df['Break_Out_Category'].value_counts()

3    20800
1    10400
0     6240
2     5200
Name: Break_Out_Category, dtype: int64

In [21]:
df.head(5)

Year LocationAbbr   LocationDesc DataSource  PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare              1          None   
1  2005           US  United States   Medicare              1          None   
2  2007           US  United States   Medicare              1          None   
3  2008           US  United States   Medicare              1          None   
4  2004           US  United States   Medicare              1          None   

   PriorityArea3 PriorityArea4                 Category  Topic  ...  \
0              1          None  Cardiovascular Diseases      3  ...   
1              1          None  Cardiovascular Diseases      3  ...   
2              1          None  Cardiovascular Diseases      1  ...   
3              1          None  Cardiovascular Diseases      1  ...   
4              1          None  Cardiovascular Diseases      3  ...   

  Break_Out_Category Break_Out  CategoryId  TopicId  IndicatorID  \
0                  3     Other          C1        4        MD501   
1                  3     Other          C1        4        MD501   
2                  0       65+          C1        3        MD301   
3                  1    Female          C1        3        MD301   
4                  2   Overall          C1        4        MD501   

   Data_Value_TypeID  BreakOutCategoryId  BreakOutId LocationID  \
0              Crude               BOC04           8         59   
1              Crude               BOC04           8         59   
2              Crude               BOC03           0         59   
3              Crude               BOC02           3         59   
4              Crude               BOC01           4         59   

  Break_Out_CategoryId  
0                    3  
1                    3  
2                    2  
3                    1  
4                    0  

[5 rows x 27 columns]

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42640 entries, 0 to 42639
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  42640 non-null  int64  
 1   LocationAbbr          42640 non-null  object 
 2   LocationDesc          42640 non-null  object 
 3   DataSource            42640 non-null  object 
 4   PriorityArea1         42640 non-null  int32  
 5   PriorityArea2         42640 non-null  object 
 6   PriorityArea3         42640 non-null  int32  
 7   PriorityArea4         42640 non-null  object 
 8   Category              42640 non-null  object 
 9   Topic                 42640 non-null  int32  
 10  Indicator             42640 non-null  object 
 11  Data_Value_Type       42640 non-null  object 
 12  Data_Value_Unit       42640 non-null  int32  
 13  Data_Value            42640 non-null  float64
 14  Data_Value_Alt        42640 non-null  float64
 15  LowConfidenceLimit 

In [23]:
df=pd.get_dummies(df,drop_first=True)
df.head()

Year  PriorityArea1  PriorityArea3  Topic  Data_Value_Unit  Data_Value  \
0  2006              1              1      3                0         5.2   
1  2005              1              1      3                0         5.1   
2  2007              1              1      1                0         8.5   
3  2008              1              1      1                0         5.1   
4  2004              1              1      3                0         6.1   

   Data_Value_Alt  LowConfidenceLimit  HighConfidenceLimit  \
0             5.2                 5.1                  5.3   
1             5.1                 5.1                  5.2   
2             8.5                 8.4                  8.5   
3             5.1                 5.1                  5.1   
4             6.1                 6.1                  6.1   

   Break_Out_Category  ...  IndicatorID_MD401  IndicatorID_MD501  \
0                   3  ...                  0                  1   
1                   3  ...                  0                  1   
2                   0  ...                  0                  0   
3                   1  ...                  0                  0   
4                   2  ...                  0                  1   

   IndicatorID_MD601  IndicatorID_MDP001  IndicatorID_MDP002  \
0                  0                   0                   0   
1                  0                   0                   0   
2                  0                   0                   0   
3                  0                   0                   0   
4                  0                   0                   0   

   IndicatorID_MDP003  IndicatorID_MDP004  BreakOutCategoryId_BOC02  \
0                   0                   0                         0   
1                   0                   0                         0   
2                   0                   0                         0   
3                   0                   0                         1   
4                   0                   0                         0   

   BreakOutCategoryId_BOC03  BreakOutCategoryId_BOC04  
0                         0                         1  
1                         0                         1  
2                         1                         0  
3                         0                         0  
4                         0                         0  

[5 rows x 145 columns]

In [24]:
x=df.drop('Topic',axis=1)
y=df['Topic']

In [25]:
x.head()

Year  PriorityArea1  PriorityArea3  Data_Value_Unit  Data_Value  \
0  2006              1              1                0         5.2   
1  2005              1              1                0         5.1   
2  2007              1              1                0         8.5   
3  2008              1              1                0         5.1   
4  2004              1              1                0         6.1   

   Data_Value_Alt  LowConfidenceLimit  HighConfidenceLimit  \
0             5.2                 5.1                  5.3   
1             5.1                 5.1                  5.2   
2             8.5                 8.4                  8.5   
3             5.1                 5.1                  5.1   
4             6.1                 6.1                  6.1   

   Break_Out_Category  TopicId  ...  IndicatorID_MD401  IndicatorID_MD501  \
0                   3        4  ...                  0                  1   
1                   3        4  ...                  0                  1   
2                   0        3  ...                  0                  0   
3                   1        3  ...                  0                  0   
4                   2        4  ...                  0                  1   

   IndicatorID_MD601  IndicatorID_MDP001  IndicatorID_MDP002  \
0                  0                   0                   0   
1                  0                   0                   0   
2                  0                   0                   0   
3                  0                   0                   0   
4                  0                   0                   0   

   IndicatorID_MDP003  IndicatorID_MDP004  BreakOutCategoryId_BOC02  \
0                   0                   0                         0   
1                   0                   0                         0   
2                   0                   0                         0   
3                   0                   0                         1   
4                   0                   0                         0   

   BreakOutCategoryId_BOC03  BreakOutCategoryId_BOC04  
0                         0                         1  
1                         0                         1  
2                         1                         0  
3                         0                         0  
4                         0                         0  

[5 rows x 144 columns]

In [26]:
y.head()

0    3
1    3
2    1
3    1
4    3
Name: Topic, dtype: int32

In [27]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [28]:
from sklearn.preprocessing import StandardScaler
scaler= StandardScaler()

x_train=scaler.fit_transform(x_train)
x_test=scaler.transform(x_test)

# Logistic Regression

In [29]:
from sklearn.linear_model import LogisticRegression
LR_model=LogisticRegression()
LR_model.fit(x_train,y_train)

LogisticRegression()

In [30]:
train_pred=LR_model.predict(x_train)
test_pred=LR_model.predict(x_test)

In [31]:
print('Train accuracy:',LR_model.score(x_train,y_train))
print('Test accuracy:',LR_model.score(x_test,y_test))


from sklearn.model_selection import cross_val_score
scores=cross_val_score(LR_model,x,y,cv=5)
print('Cv score:',scores.mean())

Train accuracy: 1.0
Test accuracy: 1.0
Cv score: 0.7573170731707318


| MODEL                | Train Score | Test Score | CV  Score |
| ---------------------| ------------| -----------|-----------|
| LOGISTIC REGRESSION  | 1.0         |       1.0  |   0.757   |

# Decision Tree

In [32]:
from sklearn.tree import DecisionTreeClassifier
dt_default=DecisionTreeClassifier()
dt_default.fit(x_train,y_train)


pred_train = dt_default.predict(x_train)
base_pred = dt_default.predict(x_test)


from sklearn.metrics import accuracy_score
print('Train accuracy:',accuracy_score(pred_train,y_train))
print('Test_accuracy:',accuracy_score(base_pred,y_test))

from sklearn.model_selection import cross_val_score

scores= cross_val_score(dt_default,x,y,cv=5)
print('cross validation score:',scores.mean())

Train accuracy: 1.0
Test_accuracy: 1.0
cross validation score: 1.0


| MODEL                | Train Score | Test Score | CV  Score |
| ---------------------| ------------| -----------|-----------|
| LOGISTIC REGRESSION  | 1.0         |       1.0  |   0.757   |
| DECISION TREE        | 1.0         |       1.0  |    1.0    |

# KNN Regression

In [33]:
from sklearn.neighbors import KNeighborsClassifier
KNN_model=KNeighborsClassifier()
KNN_model.fit(x_train,y_train)

train_pred=KNN_model.predict(x_train)
test_pred=KNN_model.predict(x_test)


In [34]:
print('Train accuracy:',KNN_model.score(x_train,y_train))
print('Test accuracy:',KNN_model.score(x_test,y_test))


from sklearn.model_selection import cross_val_score
scores=cross_val_score(KNN_model,x,y,cv=5)
print('Cv score:',scores.mean())

Train accuracy: 1.0
Test accuracy: 1.0
Cv score: 0.8765243902439025


| MODEL                | Train Score | Test Score | CV  Score |
| ---------------------| ------------| -----------|-----------|
| LOGISTIC REGRESSION  | 1.0         |       1.0  |   0.757   |
| DECISION TREE        | 1.0         |       1.0  |    1.0    |
| KNN REGRESSION       | 1.0         |       1.0  |    0.87   |

# SVM

In [35]:
from sklearn.svm import SVC
svm_model = SVC()
svm_model.fit(x_train,y_train)


train_pred=svm_model.predict(x_train)
test_pred=svm_model.predict(x_test)


In [ ]:
print('Train accuracy:',svm_model.score(x_train,y_train))
print('Test accuracy:',svm_model.score(x_test,y_test))


from sklearn.model_selection import cross_val_score
scores=cross_val_score(svm_model,x,y,cv=5)
print('Cv score:',scores.mean())

| MODEL                | Train Score | Test Score | CV  Score |
| ---------------------| ------------| -----------|-----------|
| LOGISTIC REGRESSION  | 1.0         |       1.0  |   0.757   |
| DECISION TREE        | 1.0         |       1.0  |    1.0    |
| KNN                  | 1.0         |       1.0  |    0.87   |
| SVM                  | 1.0         |       1.0  |    1      |

 # BEST MODEL = DECISION TREE